akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -77,32 +120,107 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with GenericOut
+
+      dispatchNextChunk()
+    }
+    def onRstStreamFrame(rst: RstStreamFrame): Unit = {
+      outlet.fail(new PeerClosedStreamException(rst.streamId, rst.errorCode))
+      buffer = ByteString.empty
+      wasClosed = true
+    }
+
+    private def dispatchNextChunk(): Unit = {
+      if (buffer.nonEmpty && outlet.isAvailable) {
+        val dataSize = buffer.size min maximumRequestEntityChunkSize
+        outlet.push(buffer.take(dataSize))
+        buffer = buffer.drop(dataSize)
+
+        totalBufferedData -= dataSize
+
+        log.debug(s"Dispatched chunk of $dataSize for stream [$streamId], remaining window space now $currentWindowSpace, buffered: ${buffer.size}")
+        updateWindows()
+      }
+      if (buffer.isEmpty && wasClosed) outlet.complete()
+    }
+
+    private def updateWindows(): Unit = {
+      val WindowIncrements(connectionLevel, streamLevel) = flowController.onStreamDataDispatched(
+        outstandingConnectionLevelWindow, totalBufferedData,
+        currentWindowSpace, buffer.size)
+
+      if (connectionLevel > 0) {
+        multiplexer.pushControlFrame(WindowUpdateFrame(Http2Protocol.NoStreamId, connectionLevel))
+        outstandingConnectionLevelWindow += connectionLevel
+      }
+      if (streamLevel > 0 && !wasClosed /* No reason to update window on closed streams */ ) {
+        multiplexer.pushControlFrame(WindowUpdateFrame(streamId, streamLevel))
+        currentWindowSpace += streamLevel
+      }
+
+      log.debug(
+        s"adjusting con-level window by $connectionLevel, stream-level window by $streamLevel," +
+          s"remaining window space now $currentWindowSpace, buffered: ${buffer.size}," +
+          s"remaining connection window space now $outstandingConnectionLevelWindow, total buffered: $totalBufferedData")
+    }
+  }
+
+  sealed abstract class ReceivingData(afterEndStreamReceived: IncomingStreamState) extends IncomingStreamState { _: Product ⇒
+    protected def buffer: Buffer
     def handle(event: StreamFrameEvent): IncomingStreamState = event match {
       case d: DataFrame ⇒
-        outlet.push(d.payload)
+        outstandingConnectionLevelWindow -= d.sizeInWindow
+        totalBufferedData += d.payload.size // padding can be seen as instantly discarded
+
+        // FIXME: if (incomingConnectionLevelWindow < 0) fail the connection with FLOW_CONTROL_ERROR
+
+        val windowSizeIncrement = flowController.onConnectionDataReceived(outstandingConnectionLevelWindow, totalBufferedData)
+        if (windowSizeIncrement > 0) {
+          multiplexer.pushControlFrame(WindowUpdateFrame(Http2Protocol.NoStreamId, windowSizeIncrement))
+          outstandingConnectionLevelWindow += windowSizeIncrement
+        }
+
+        buffer.onDataFrame(d)
         maybeFinishStream(d.endStream)
       case r: RstStreamFrame ⇒
-        outlet.fail(new PeerClosedStreamException(r.streamId, r.errorCode))
+        buffer.onRstStreamFrame(r)
```

Comment:
btw; love the general move to these named on* callbacks (everywhere in the infra) 👍